frame/node-authorization/src/lib.rs

```diff
@@ -0,0 +1,761 @@
+        /// The given node was claimed by a user.
+        NodeClaimed(NodePublicKey, AccountId),
+        /// The given claim was removed by its owner.
+        ClaimRemoved(NodePublicKey, AccountId),
+        /// The node was transferred to another account.
+        NodeTransferred(NodePublicKey, AccountId, AccountId),
+        /// The allowed connections were added to a node.
+        ConnectionsAdded(NodePublicKey, Vec<NodePublicKey>),
+        /// The allowed connections were removed from a node.
+        ConnectionsRemoved(NodePublicKey, Vec<NodePublicKey>),
+    }
+}
+
+decl_error! {
+    /// Error for the node authorization module.
+    pub enum Error for Module<T: Trait> {
+        /// Too many well known nodes.
+        TooManyNodes,
+        /// The node is already joined in the list.
+        AlreadyJoined,
+        /// The node doesn't exist in the list.
+        NotExist,
+        /// The node is already claimed by a user.
+        AlreadyClaimed,
+        /// The node hasn't been claimed yet.
+        NotClaimed,
+        /// You are not the owner of the node.
+        NotOwner,
+        /// No permisson to perform specific operation.
+        PermissionDenied,
+    }
+}
+
+decl_module! {
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+        /// The maximum number of authorized well known nodes
+        const MaxWellKnownNodes: u32 = T::MaxWellKnownNodes::get();
+
+        type Error = Error<T>;
+
+        fn deposit_event() = default;
+
+        /// Add a node to the set of well known nodes. If the node is already claimed, the owner
+        /// will be updated and keep the existing additional connection unchanged.
+        ///
+        /// May only be called from `T::AddOrigin`.
+        ///
+        /// - `node`: identifier of the node.
+        #[weight = (T::WeightInfo::add_well_known_node(), DispatchClass::Operational)]
+        pub fn add_well_known_node(origin, node: NodePublicKey, owner: T::AccountId) {
+            T::AddOrigin::ensure_origin(origin)?;
+
+            let mut nodes = WellKnownNodes::get();
+            ensure!(nodes.len() < T::MaxWellKnownNodes::get() as usize, Error::<T>::TooManyNodes);
+
+            let location = nodes.binary_search(&node).err().ok_or(Error::<T>::AlreadyJoined)?;
+            nodes.insert(location, node.clone());
+            
+            WellKnownNodes::put(&nodes);
+            <Owners<T>>::insert(&node, &owner);
+
+            Self::deposit_event(RawEvent::NodeAdded(node, owner));
+        }
+
+        /// Remove a node from the set of well known nodes. The ownership and additional
+        /// connections of the node will also be removed.
+        ///
+        /// May only be called from `T::RemoveOrigin`.
+        ///
+        /// - `node`: identifier of the node.
+        #[weight = (T::WeightInfo::remove_well_known_node(), DispatchClass::Operational)]
+        pub fn remove_well_known_node(origin, node: NodePublicKey) {
```

Comment:
I think the admin-level API to alter `Owners/AdditioanlConnections` for nodes that are not `WellKnownNodes` is missing.

Reply:
There are general apis to `remove_claim`, `transfer_node`, `add` and `remove_connections`. I made these not admin api because I think it's the opinions of node operators, not something the overall groups of the chain should cared about.

Reply:
The state of the contract can grow uncontrollably, there for sure might be cases where the owner keys are going to get lost, having an admin API to clean up might really be needed. That said, it's always possible to change the storage via governance, so maybe I'm exaggerating.

Reply:
I assume there will be a customized governance mechanism similar to sudo pallet which can provide a `sudo_as` call to do such cleaning, It's like general requirements for functionalities provided by a permissioned blockchain. I'm also interested how the API gonna change when more use cases comes in.